Include page for cross compile to an old x86 (#17)
epics-docs/how-tos: building-epics/cross-compile-epics-and-a-ioc-to-an-old-x86-linux.md

## Overview

I’m assuming you, like me, are running Ubuntu 64 bits. I’m also assuming you already have EPICS base downloaded and compiled. We will go through the steps of downloading, configuring and compiling [Crosstool-NG](http://crosstool-ng.org/), which will be used to generate our toolchain. Then we will download and compile a couple of libraries needed by EPICS (namely [readline](https://www.gnu.org/software/readline/) and [ncurses](https://www.gnu.org/software/ncurses/)). Finally, we will compile EPICS base and an example IOC to our target using the newly built toolchain.

## Crosstool-NG

### Downloading and extracting

First we get the tarball containing the source code and extract it.

``` console
wget http://crosstool-ng.org/download/crosstool-ng/crosstool-ng-1.9.3.tar.bz2
tar -xvf crosstool-ng-1.9.3.tar.bz2
```

Crosstool-NG has a lot of dependencies, you might want to get them before compiling:

``` console
sudo apt-get install gawk bison flex texinfo automake libtool cvs libcurses5-dev build-essential
```

### Compiling

In order to compile:

> ``` console

Reviewer: If this is a script without a `$`, this should be:

```suggestion
``` bash
```

cd crosstool-ng-1.9.3
./configure --local
make
```

### Configuring

Before configuring Crosstool-NG, I gathered information about my target system:

Kernel Version:

``` console
$ uname -r
2.6.27.27
```

GCC Version:

``` console
$ gcc --version
gcc (GCC) 4.2.4
```

glibc version:

``` console 
$ /lib/libc.so.6
GNU C Library stable release version 2.7, by Roland McGrath et al.
```

binutils version:

``` console
$ ld --version
GNU ld (Linux/GNU Binutils) 2.18.50.0.9.20080822
```

Based on this information, and on a lot of trial and error, I configured Crosstool-NG as follows (everything else set as default):

``` console

> ./ct-ng menuconfig

Reviewer: ```suggestion
$ ./ct-ng menuconfig
```

PATHS AND MISC OPTIONS
   [*] Use obsolete features
   
TARGET OPTIONS
   Target architecture (x86)
   Bitness: (32-bit)
   (i686) Architecture Level
   
OPERATING SYSTEM
   Target OS (linux)
   Linux kernel version (2.6.27.55)
   
BINARY UTILITIES
   Binutils version (2.17)
   
C-COMPILER
   GCC version (4.2.4)
   [*] C++
   
C-LIBRARY
   glibc version (2.6.1)
   
```

I tried other configurations, but they crashed the compilation process.

### Compiling the toolchain

Now we can compile the toolchain:

> ```

Reviewer: ```suggestion
``` bash
```

./ct-ng build
```

This will take a while. Go get some coffee or watch a cat video on Youtube.

Once built, the toolchain will be in `$HOME/x-tools/i686-unknown-linux-gnu/`

It’s a good idea (I think) to put the cross-compiler binaries on your path. Add this to the end of your ~/.bashrc:

> ```

Reviewer: ```suggestion
``` bash
```

PATH=$PATH:$HOME/x-tools/i686-unknown-linux-gnu/bin
```

Then source your .bashrc, so the changes take effect.

> ```

Reviewer: ```suggestion
``` bash
```

. ~/.bashrc
```

## EPICS dependencies

In order to properly build epics-base to our target system, we have to take care of EPICS dependencies first. Namely, the libraries ‘readline’ and ‘ncurses’.

They will be installed in our toolchain directory. We have to make it writable:

``` console
chmod -R +w $HOME/x-tools/i686-unknown-linux-gnu/i686-unknown-linux-gnu/sys-root/usr
```

Now we can download, configure, compile and install the libraries.

### readline

> ``` console

Reviewer: If think those two should be `bash` also

### Recompile EPICS base

Now, we recompile EPICS base:

``` console
make clean uninstall
make
```

Hopefully, we have everything in place to start developing our IOC’s.

## Example IOC

Let’s create a working directory for our programs. I decided to call it ‘apps’:

``` console
mkdir ~/apps
```

### Creating

To create an example IOC, we use a Perl script provided by EPICS:

``` console
cd ~/apps
mkdir myexample
cd myexample
makeBaseApp.pl -t example myexample
```

The last command tells the script to create an application named ‘myexample’ using the template (option -t) ‘example’. Now we make our IOC bootable

``` console
makeBaseApp.pl -i -t example myexample
```

It will ask you what the target is. We went to all this trouble to be able to select:

``` console
linux-686
```

For the Application Name, just hit enter.

### Configuring

Add this line to the file `~/apps/myexample/configure/CONFIG_SITE`

``` makefile
STATIC_BUILD=YES
```

This will statically link EPICS libraries into our executable.

Now, let’s consider that you will put your IOC in the folder /home of your target system. Edit the file ~/apps/myexample/iocBoot/iocmyexample/envPaths, so it will be:

> ``` bash

Reviewer: I think you'll get better syntax highlighting by using `csh`

```suggestion
``` csh
```

The first approach is the easiest, but compiling from inside a VM can be slow and the distribution was not very user friendly. So I took the second path, which I’ll describe in this document.

## Overview

I’m assuming you, like me, are running Ubuntu 64 bits. I’m also assuming you already have EPICS base downloaded and compiled. We will go through the steps of downloading, configuring and compiling [Crosstool-NG](http://crosstool-ng.org/), which will be used to generate our toolchain. Then we will download and compile a couple of libraries needed by EPICS (namely [readline](https://www.gnu.org/software/readline/) and [ncurses](https://www.gnu.org/software/ncurses/)). Finally, we will compile EPICS base and an example IOC to our target using the newly built toolchain.

## Crosstool-NG

### Downloading and extracting

First we get the tarball containing the source code and extract it.

```
wget http://crosstool-ng.org/download/crosstool-ng/crosstool-ng-1.9.3.tar.bz2
tar -xvf crosstool-ng-1.9.3.tar.bz2
```

Crosstool-NG has a lot of dependencies, you might want to get them before compiling:

> sudo apt-get install gawk bison flex texinfo automake libtool cvs libcurses5-dev build-essential

Reviewer 2: This should be a code block as well

### Compiling

In order to compile:

```

> cd crosstool-ng-1.9.3

Reviewer 3: I think that you should try be more consistent about commands starting with `$`. There is the option to use the `console` syntax highlighting scheme.

Author: The ones with $ are the ones that the result of the command is showed (I suppose), is that too bad ?

Author: using bash syntax

./configure --local
make
```

### Configuring

Before configuring Crosstool-NG, I gathered information about my target system:

Kernel Version:

```

> $uname -r

Reviewer 3: Similarly here to the above. This should be `$ uname -r` (with a space), and use `console` syntax highlighting.

Author: Here in this case, the $ is used to indicate what is the command and what is the result (I suppose, as this was copied)

Author: Using bash syntax for when no $ as suggested by @minijackson

2.6.27.27
```

GCC Version:

```

> $gcc --version

Reviewer 3: etc.

gcc (GCC) 4.2.4
```

glibc version:

```
$/lib/libc.so.6
GNU C Library stable release version 2.7, by Roland McGrath et al.
```

binutils version:

```
$ld --version
GNU ld (Linux/GNU Binutils) 2.18.50.0.9.20080822
```

Based on this information, and on a lot of trial and error, I configured Crosstool-NG as follows (everything else set as default):

```
./ct-ng menuconfig

PATHS AND MISC OPTIONS

>    \[\*\] Use obsolete features

Reviewer 3: Are these escape characters correct? (and below, see C-COMPILER)

Author: Nope, fixing.

   
TARGET OPTIONS
   Target architecture (x86)
   Bitness: (32-bit)
   (i686) Architecture Level
   
OPERATING SYSTEM
   Target OS (linux)
   Linux kernel version (2.6.27.55)
   
BINARY UTILITIES
   Binutils version (2.17)
   
C-COMPILER
   GCC version (4.2.4)
   \[\*\] C++
   
C-LIBRARY
   glibc version (2.6.1)
   
```

I tried other configurations, but they crashed the compilation process.

### Compiling the toolchain

Now we can compile the toolchain:

```
./ct-ng build
```

This will take a while. Go get some coffee or watch a cat video on Youtube.

> Once built, the toolchain will be in $HOME/x-tools/i686-unknown-linux-gnu/

Reviewer 2: I would add the $HOME/x-tools/i686-unknown-linux-gnu/ into code blocks as well (single quote)

PATH=$PATH:$HOME/x-tools/i686-unknown-linux-gnu/bin
```

Then source your .bashrc, so the changes take effect.

```
. ~/.bashrc
```

## EPICS dependencies

In order to properly build epics-base to our target system, we have to take care of EPICS dependencies first. Namely, the libraries ‘readline’ and ‘ncurses’.

They will be installed in our toolchain directory. We have to make it writable:

```
chmod -R +w $HOME/x-tools/i686-unknown-linux-gnu/i686-unknown-linux-gnu/sys-root/usr
```

Now we can download, configure, compile and install the libraries.

### readline

```
wget ftp://ftp.cwru.edu/pub/bash/readline-6.2.tar.gz
tar -xvf readline-6.2.tar.gz
cd readline-6.2
./configure --prefix=$HOME/x-tools/i686-unknown-linux-gnu/i686-unknown-linux-gnu/sys-root/usr --host=i686-unknown-linux-gnu
make
make install
```

### ncurses

```
wget ftp://ftp.gnu.org/pub/gnu/ncurses/ncurses-5.9.tar.gz
tar -xvf ncurses-5.9.tar.gz
cd ncurses-5.9
./configure --prefix=$HOME/x-tools/i686-unknown-linux-gnu/i686-unknown-linux-gnu/sys-root/usr --host=i686-unknown-linux-gnu
make
make install
```

## Configure cross-compilation in EPICS

We’re almost done. Back to the epics-base directory, open the file:

```

> $EPICS\_BASE/configure/CONFIG\_SITE

Reviewer 3: Here is an example of incorrect escapes: `EPICS_BASE` vs `EPICS\_BASE`. Could also be on the same line...

```

Change the line:

```

> CROSS\_COMPILER\_TARGET\_ARCHS=

Reviewer 3: More escapes. This should probably be `makefile` syntax highlighgting.

```

To:

```

> CROSS\_COMPILER\_TARGET\_ARCHS=linux-686

Reviewer 3: escape

```

This tells EPICS base to be compiled both for the host system and for the linux-686 target.

Save and close. Now we will create our own target configuration file, based on a existing file.

```
cd $EPICS\_BASE/configure/os

> cp CONFIG\_SITE.Common.linux-x86 CONFIG\_SITE.Common.linux-686

Reviewer 3: escapes

```

> Open CONFIG\_SITE.Common.linux-686 and edit it. Comment out the line:

Reviewer 3: This path should be monotype, and should have its escape characters fixed.

```

> #COMMANDLINE\_LIBRARY = READLINE

Reviewer 3: makefile highlighting, escapes

```

And uncomment:

```

> COMMANDLINE\_LIBRARY = READLINE\_NCURSES

Reviewer 3: highlighting, escapes

```

At the end of the file, add the lines:

```
GNU\_DIR=$HOME/x-tools/i686-unknown-linux-gnu

> GNU\_TARGET=i686-unknown-linux-gnu

Reviewer 3: highlighting, escapes

```

This tells EPICS to search for both readline and ncurses libraries during compilation. The last two lines indicate the location of the toolchain and its prefix. Save and close. Last file to edit: CONFIG.Common.linux-686

Change the line that says

```
VALID\_BUILDS = Ioc
```

To

```

> VALID\_BUILDS = Host Ioc

Reviewer 3: highlighting, escapes

```

This is needed in order to get caRepeater compiled, according to [this source](https://epics.anl.gov/tech-talk/2012/msg01102.php).

### Recompile EPICS base

Now, we recompile EPICS base:

```
make clean uninstall
make
```

Hopefully, we have everything in place to start developing our IOC’s.

## Example IOC

Let’s create a working directory for our programs. I decided to call it ‘apps’:

```
mkdir ~/apps
```

### Creating

To create an example IOC, we use a Perl script provided by EPICS:

```
cd ~/apps
mkdir myexample
cd myexample
makeBaseApp.pl -t example myexample
```

The last command tells the script to create an application named ‘myexample’ using the template (option -t) ‘example’. Now we make our IOC bootable

```
makeBaseApp.pl -i -t example myexample
```

It will ask you what the target is. We went to all this trouble to be able to select:

```
linux-686
```

For the Application Name, just hit enter.

### Configuring

> Add this line to the file ~/apps/myexample/configure/CONFIG\_SITE

Reviewer 3: monotype file name

```

> STATIC\_BUILD=YES

Reviewer 3: escapes, highlighting

```

This will statically link EPICS libraries into our executable.

> Now, let’s consider that you will put your IOC in the folder /home of your target system. Edit the file ~/apps/myexample/iocBoot/iocmyexample/envPaths, so it will be:

Reviewer 3: monotype file name

```
epicsEnvSet("ARCH","linux-686")
epicsEnvSet("IOC","iocskel")
epicsEnvSet("TOP","/home/myexample")

> epicsEnvSet("EPICS\_BASE","/home/myexample")

Reviewer 3: Oooh, sneaky escape in there.

```

Note that we set EPICS base to coincide with the <top> folder of our IOC. I did it because the IOC depends on the caRepeater program, which would be present in an EPICS base if we had one in our target. Because we don’t, I’ll simply copy the caRepeater generated by the host to the the /bin folder of our IOC <top> folder:

```

> cp $EPICS\_BASE/bin/linux-686/caRepeater ~/apps/myexample/bin/linux-686/

Reviewer 3: escapes (where on earth did these all come from?)

Author: sorry :D I think is because I used a site to convert html to md, and it make these all escapes. I will double check on next PRs.

```

### Compiling

Compile the IOC and prepare it for execution.

```
make
chmod +x iocBoot/iocmyexample/st.cmd
```

Note that you won’t be able to run the IOC in your host system, given that it was compiled to our target system. You won’t be able to run it in your target system neither, as your target lacks three libraries: two needed by caRepeater and one needed by the IOC.

First, take care of the libraries needed by caRepeater:

```
mkdir ~/apps/myexample/lib
mkdir ~/apps/myexample/lib/linux-686/
cp $EPICS\_BASE/lib/linux-686/libca.so.3.14 ~/apps/myexample/lib/linux-686

> cp $EPICS\_BASE/lib/linux-686/libCom.so.3.14 ~/apps/myexample/lib/linux-686

Reviewer 3: escapes

# How To cross compile EPICS and a IOC to an old x86 Linux system

## Introduction

I was given the task of developing a IOC which should run in a x86 PC with an old Linux distribution. My development machine was a x86_64 PC running Ubuntu 12.04. EPICS does a great job cross compiling from a 64-bits host to a 32-bits target if both have compatible versions of glibc, binutils, kernel, etc. In my case, however, my target had much older versions. I considered two different solutions:

1.  Create a Virtual Machine and install the target’s Linux distribution. From the Virtual Machine, compile EPICS and my IOC, and then run the IOC in the target
2.  Build a toolchain configured for my target and use that toolchain to compile both EPICS and the IOC.

The first approach is the easiest, but compiling from inside a VM can be slow and the distribution was not very user friendly. So I took the second path, which I’ll describe in this document.

## Overview

> I’m assuming you, like me, are running Ubuntu 64 bits. I’m also assuming you already have EPICS base downloaded and compiled. We will go through the steps of downloading, configuring and compiling [Crosstool-NG](http://crosstool-ng.org/), which will be used to generate our toolchain. Then we will download and compile a couple of libraries needed by EPICS (namely [readline](https://www.gnu.org/software/readline/) and [ncurses](https://www.gnu.org/software/ncurses/)). Finally, we will compile EPICS base and an example IOC to our target using the newly built toolchain.

Reviewer: Very nitpicky, but this might cause encoding issues between Windows/Linux/macOS

```suggestion
I'm assuming you, like me, are running Ubuntu 64 bits. I'm also assuming you already have EPICS base downloaded and compiled. We will go through the steps of downloading, configuring and compiling [Crosstool-NG](http://crosstool-ng.org/), which will be used to generate our toolchain. Then we will download and compile a couple of libraries needed by EPICS (namely [readline](https://www.gnu.org/software/readline/) and [ncurses](https://www.gnu.org/software/ncurses/)). Finally, we will compile EPICS base and an example IOC to our target using the newly built toolchain.
```

# How To cross compile EPICS and a IOC to an old x86 Linux system

## Introduction

> I was given the task of developing a IOC which should run in a x86 PC with an old Linux distribution.
> My development machine was a x86_64 PC running Ubuntu 12.04.
> EPICS does a great job cross compiling from a 64-bits host to a 32-bits target if both have compatible versions of glibc,
> binutils,
> kernel,
> etc.
> In my case,
> however,
> my target had much older versions. I considered two different solutions:

Reviewer: I would write it like that, but of course there's a bit of interpretation:

```suggestion
I was given the task of developing a IOC
which should run in a x86 PC with an old Linux distribution.
My development machine was a x86_64 PC running Ubuntu 12.04.
EPICS does a great job cross compiling from a 64-bits host to a 32-bits target
if both have compatible versions of glibc, binutils, kernel, etc.
In my case, however,
my target had much older versions.
I considered two different solutions:
```

I *think* those `which` and `if` introduce a relative clause, so it's good to have a line break before. If I'm note sure, I just usually cheat and add a comma.

> 1.  Create a Virtual Machine and install the target's Linux distribution. From the Virtual Machine,
> compile EPICS and my IOC,
> and then run the IOC in the target
> 2.  Build a toolchain configured for my target and use that toolchain to compile both EPICS and the IOC.

Reviewer: You can indent the numbered list (although it's not specified in the style guide, so either is fine).

```suggestion
1.  Create a Virtual Machine and install the target's Linux distribution.
    From the Virtual Machine,
    compile EPICS and my IOC,
    and then run the IOC in the target
2.  Build a toolchain configured for my target
    and use that toolchain to compile both EPICS and the IOC.
```

The first approach is the easiest,

> but compiling from inside a VM can be slow and the distribution was not very user friendly.

Reviewer: I think you can break before `and` again. Rest is good :+1: 

```suggestion
but compiling from inside a VM can be slow
and the distribution was not very user friendly.
```
